src/aws-route53resolver/resolver-rule.ts

```diff
@@ -0,0 +1,170 @@
+import { Resource, Token } from 'aws-cdk-lib';
+import { IVpc } from 'aws-cdk-lib/aws-ec2';
+import {
+  CfnResolverRule,
+  CfnResolverRuleAssociation,
+} from 'aws-cdk-lib/aws-route53resolver';
+import { Construct } from 'constructs';
+import { Protocol } from './protocol';
+import { IResolverEndpoint } from './resolver-endpoint';
+
+/**
+ * A Resolver Rule.
+ */
+export interface IResolverRule {
+  readonly resolverRuleId: string;
+}
+
+enum RuleType {
+  FORWARD = 'FORWARD',
+  SYSTEM = 'SYSTEM',
+  RECURSIVE = 'RECURSIVE',
+}
+
+/**
+ * The target of the Resolver Rule.
+ */
+export interface Target {
+  /**
+   * The IPv4 Address of the target.
+   */
+  readonly ipv4?: string;
+  /**
+   * The IPv6 Address of the target.
+   */
+  readonly ipv6?: string;
+  /**
+   * The target port.
+   *
+   * @default 53
+   */
+  readonly port?: number;
+  /**
+   * The transmission protocol.
+   *
+   * @default Protocol.DO_53
+   */
+  readonly protocol?: Protocol;
+}
+
+/**
+ * Options for a Resolver Rule.
+ */
+interface ResolverRuleOptions {
+  /**
+   * The name of the domain for which you want to forward queries.
+   */
+  readonly domainName: string;
+}
+
+interface ResolverRuleBaseProps extends ResolverRuleOptions {
+  readonly vpcs: IVpc[];
+  readonly outboundEndpoint?: IResolverEndpoint;
+  readonly targets?: Target[];
+  readonly ruleType: RuleType;
+}
+
+class ResolverRuleBase extends Resource implements IResolverRule {
+  readonly resolverRuleId: string;
+  constructor(scope: Construct, id: string, props: ResolverRuleBaseProps) {
+    super(scope, id);
+    const resource = new CfnResolverRule(this, 'Resource', {
```

Comment:
Please extract the creation of sub-constructs into a factory method that can be overridden in a subclass